{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"package 'igraph' was built under R version 3.6.3\"\n",
      "Attaching package: 'igraph'\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    union\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggtree' was built under R version 3.6.3\"Registered S3 methods overwritten by 'ggplot2':\n",
      "  method         from \n",
      "  [.quosures     rlang\n",
      "  c.quosures     rlang\n",
      "  print.quosures rlang\n",
      "Registered S3 method overwritten by 'treeio':\n",
      "  method     from\n",
      "  root.phylo ape \n",
      "ggtree v2.0.2  For help: https://yulab-smu.github.io/treedata-book/\n",
      "\n",
      "If you use ggtree in published research, please cite the most appropriate paper(s):\n",
      "\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi: 10.1093/molbev/msy194\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'phangorn' was built under R version 3.6.3\"Loading required package: ape\n",
      "Warning message:\n",
      "\"package 'ape' was built under R version 3.6.3\"\n",
      "Attaching package: 'ape'\n",
      "\n",
      "The following object is masked from 'package:ggtree':\n",
      "\n",
      "    rotate\n",
      "\n",
      "The following objects are masked from 'package:igraph':\n",
      "\n",
      "    edges, mst, ring\n",
      "\n",
      "\n",
      "Attaching package: 'phangorn'\n",
      "\n",
      "The following object is masked from 'package:igraph':\n",
      "\n",
      "    diversity\n",
      "\n",
      "treeio v1.10.0  For help: https://yulab-smu.github.io/treedata-book/\n",
      "\n",
      "If you use treeio in published research, please cite:\n",
      "\n",
      "LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package for phylogenetic tree input and output with richly annotated and associated data. Molecular Biology and Evolution 2019, accepted. doi: 10.1093/molbev/msz240\n",
      "\n",
      "\n",
      "Attaching package: 'treeio'\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    drop.tip\n",
      "\n",
      "The following object is masked from 'package:igraph':\n",
      "\n",
      "    parent\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from 'package:igraph':\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Warning message:\n",
      "\"package 'S4Vectors' was built under R version 3.6.2\"Loading required package: stats4\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following object is masked from 'package:ggtree':\n",
      "\n",
      "    expand\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Warning message:\n",
      "\"package 'IRanges' was built under R version 3.6.2\"\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "The following object is masked from 'package:ggtree':\n",
      "\n",
      "    collapse\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "The following object is masked from 'package:treeio':\n",
      "\n",
      "    mask\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    complement\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "Warning message:\n",
      "\"package 'insect' was built under R version 3.6.3\"Registered S3 method overwritten by 'openssl':\n",
      "  method      from\n",
      "  print.bytes Rcpp\n",
      "\n",
      "Attaching package: 'insect'\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    trim\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "The following object is masked from 'package:ggtree':\n",
      "\n",
      "    expand\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"igraph\")\n",
    "library(\"ggtree\")\n",
    "library(\"phangorn\")\n",
    "library(\"treeio\")\n",
    "library(\"Biostrings\")\n",
    "library(\"msa\")\n",
    "library(\"ape\")\n",
    "library(\"insect\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Построем филогенетическое дерево по геномам различных штаммов коронавируса. Данные загружены с сайта [NCBI](https://www.ncbi.nlm.nih.gov/genbank/sars-cov-2-seqs/#nucleotide-sequences) и собраны в один файл `common.fasta` в формате `fasta`. Длины ДНК-последовательностей отличаются, поэтому предварительно требуется выровнять их. Это можно сделать с помощью следующего кода:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_data <- read.fasta(\"common.fasta\")\n",
    "l = c(dna2char(fasta_data[1]))\n",
    "for (i in 2:16) {\n",
    "    l = rbind(l, c(dna2char(fasta_data[i])))\n",
    "}\n",
    "string.set <- DNAStringSet(l)\n",
    "string.set <- msa(DNAStringSet(l))\n",
    "fasta_data <- as.DNAbin(string.set)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Демонстрация работы на маленьких данных:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "use default substitution matrix\n",
      "CLUSTAL 2.1  \n",
      "\n",
      "Call:\n",
      "   msa(DNAStringSet(c(dna2char(Q1), dna2char(Q2), dna2char(Q3))))\n",
      "\n",
      "MsaDNAMultipleAlignment with 3 rows and 14 columns\n",
      "    aln \n",
      "[1] TCCGAATAAGTAAA\n",
      "[2] -CCGAATCAGTA--\n",
      "[3] TCTAAATAAGCAC-\n",
      "Con TCCGAATAAGTA?- \n"
     ]
    }
   ],
   "source": [
    "Q1 <- as.DNAbin(c(\"T\",\"C\",\"C\",\"G\",\"A\",\"A\",\"T\",\"A\",\"A\",\"G\",\"T\",\"A\",\"A\",\"A\"))\n",
    "Q2 <- as.DNAbin(c(\"C\",\"C\",\"G\",\"A\",\"A\",\"T\",\"C\",\"A\",\"G\",\"T\",\"A\"))\n",
    "Q3 <- as.DNAbin(c(\"T\",\"C\",\"T\",\"A\",\"A\",\"A\",\"T\",\"A\",\"A\",\"G\",\"C\",\"A\",\"C\"))\n",
    "print(msa(DNAStringSet(c(dna2char(Q1),dna2char(Q2),dna2char(Q3)))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Однако для выравнивания нескольких десятков последовательностей длины порядка $30000$ нуклеотидов у меня не достаточно вычислительной мощности (это займет слишком много времени), поэтому для получения того же результата воспользуемся онлайн-ресурсом [Clustal Omega](https://www.ebi.ac.uk/Tools/msa/clustalo/). Полученный в результате файл с выровненными ДНК-последовательностями назовем `common_msa.fasta`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Теперь построем филогенетического дерева. Для удобства в подписи вершин вынесем только **id** штамма, страну и дату обнаружения."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnames <- c(\"MT007544 25/01/20 AU\",\n",
    "            \"MT039890 01/20 KR\",\n",
    "            \"MN988713 21/01/20 USA IL\",\n",
    "            \"MT291832 25/01/20 CN BJ\",\n",
    "            \"MN985325 19/01/20 USA WA\",\n",
    "            \"MN938384 10/01/20 CN SZ\",\n",
    "            \"MN997409 22/01/20 USA AZ\",\n",
    "            \"MT291830 30/12/19 CN WHN\",\n",
    "            \"LC529905 01/20 JP\",\n",
    "            \"MT019529 23/12/19 CN WHN\",\n",
    "            \"MN994468 22/01/20 USA CA\",\n",
    "            \"MT072688 13/01/20 NP\",\n",
    "            \"MT039873 20/01/20 CN HNG\",\n",
    "            \"LR757996 01/01/20 CN WHN\",\n",
    "            \"MT192772 22/01/20 VN\",\n",
    "            \"MT192759 25/01/20 TW\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_data <- as.matrix(read.fasta(\"common_msa.fasta\"))\n",
    "rownames(fasta_data) <- rnames\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
